### ISSUE TITLE
Understanding on the Road-Map on Delta Scan Feature for Scan-Code 

### ISSUE THREAD

**Issue by porsche-rishisaxena** [oss-review-toolkit/ort#4347]
Is there any road-map designed for enabling "delta-scan" feature for Scan-Code from ORT standpoint? 
If yes, please share some information.

**Comment by sschuberth** [comment]
No, a delta-scan feature for use together with ScanCode is not currently planned.

**Comment by nikpete** [comment]
@sschuberth We do have several use cases where delta scan capabilities will make our life much easier as we have to tranfer all results to a subsequent legal workflow tool. 
Is this something that was not required / requested  by the community so far and / or how do you currently handle delta scans? I just want to get a proper understanding on this topic and if we can pick it up...not sure how difficult it might be to introduce a delta-scan feature...you are the expetert :-)

**Comment by tsteenbe** [comment]
@fviernau and myself had some private discussions a few weeks ago on adding a file hash cache to enable delta scanning once scan-by-repository has been merged. Looking at ways to speed up ORT scans in CI/CD, believe cache to be cheaper, simpler and cross scanner solution compared to using [DeltaCode](https://github.com/nexB/deltacode/)

**Comment by sschuberth** [comment]
> Is this something that was not required / requested by the community so far

Indeed, this has not been requested so far. I believe we had some internal discussions about this a really long time ago, but since it's not trivial to implement we first wanted to get the "basic" mode running.

> and / or how do you currently handle delta scans?

We simple don't, and always do full scans if something changes.

**Comment by fviernau** [comment]
A very simple and probably good enough implementation could be to use a map which maps
`sha(file) -> (copyright findings, license findings)`. That map could become quite large so a LRU strategy would need to be used.
Such map could be really used as transparent cache (not a persistent storage) just for speeding things up.

The scanner implementation would need to be changes by first computing the SHA1 of all files and looking up the cache entries, the scan the remaining files. IIRC ScanCode internally has some ignore patterns, so some effort may be required to handle these properly.

**Comment by sschuberth** [comment]
For delta-scanning, I believe the tricky bit to be to determine the list of files that *do* need to be scanned, not to determine / cache the results for known files.

**Comment by rbieniek** [comment]
First of all, I would like to clarify what we at Porsche understand as a delta scan. Given our legal workflow, we want to build a full report for the first iteration of a project. For any subsequent release / revision, we want the lawyers to revisit only those dependencies which have changed. And by changed, we mean a dependency which has either been freshly introduced or updated with a more recent version.

To break it down into use cases, there are three different cases to be considered for a generic delta scan capability, given that you want to build a delta between two iterations (#1 and #2)  of the same project:
1) some new dependencies are introduced in the transition from #1 to #2
2) the set of dependencies stays the same but updated versions of dependencies have been introduced
3) dependencies are removed in the transition between #1 and #2

The use case 3) is, of course, trivial and IMHO there is nothing to be done for it.

For use cases 1) and 2) it is not that trivial. The afformentioned hash calculation can be a quick way to detect the need for a delta scan. But it doesn't solve the problem.

We built a simple test case to demonstrate our requirements and the expected result:
- The base project is a simple spring boot project with a simple set of dependencies:
- <dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-validation</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>
This set of depenencies models our baseline project andn would be passed over to the lawyer for review

- in the second iteration, we introduce just one additional dependency (spring data mongodb), thus leading to this set of dependencies:
- <dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-validation</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
		      <groupId>org.springframework.boot</groupId>
		      <artifactId>spring-boot-starter-data-mongodb</artifactId>
		    </dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>

- Our expectation is that a delta scan results consists only of the spring boot data mongodb dependency and all the additional dependencies  the spring data mongodb dependency relies on.

We have given it a try by using the orth subtract scan result function to build a difference, thus delta, between the two revisions but this didn't deliver a minimal set of add dependecies that we expected.

Implementationwise, I think ORT needs to be enhanced with a persistence capability to store the overall scan result per project and revision / timestamp of the scan, so that the difference between project iterations can be determined in the first place.
This persistence capabilty should support filesystem-based persistence as well as database- oder ReST-based persistence. And it should support an efficient way to determine the changeset between two project iterations.

**Comment by fviernau** [comment]
@rbieniek the terminology "delta scanning" (for ScanCode) was so far being used for a non-function transparent performance optimization, e.g. to be achieved by using (finer granularity) caching. So Porsche's delta scanning IMO is a different unrelated topic.

**Comment by sschuberth** [comment]
Put on hold on behalf of @nikpete.

**Comment by arieltorti** [comment]
We have the same use cases as @rbieniek and are interested in helping solve this.

Could we re-open this discussion ?

**Comment by sschuberth** [comment]
> Could we re-open this discussion ?

Sure, feel free to just type what's on your mind 😁 

But in order to continue the discussion I'd like to ensure we're all on the same page, as several terminologies were used and I fear we're not always talking about the same thing:

1. In a nutshell, Porsche wants a "diff" between original dependencies and dependencies in a later revision of the product. Eventually, the added and changed dependencies' source code should be scanned.
2. This is fundamentally different from what @tsteenbe and @fviernau were [referring](https://github.com/oss-review-toolkit/ort/issues/4347#issuecomment-896089579) to [above](https://github.com/oss-review-toolkit/ort/issues/4347#issuecomment-896097724), which is about delta-scanning on the file level of only those files that changed between the versions of a single dependency.

So @arieltorti and @dgutson, are you interested in 1. oder 2.?

**Comment by arieltorti** [comment]
If I understand it correctly I would say 2 is a non functional feature that would reduce ORT execution times.

We're interested in option 1. Having a "diff" (delta) of dependencies between two analysis at the report stage: which dependencies were updated/added/removed.
